### PR TITLE
[train] Fix regression where large Trainer attributes get serialized along with actor class

### LINKED ISSUE #43191: Release test read_images_train_4_gpu.aws failed
Release test **read_images_train_4_gpu.aws** failed. See https://buildkite.com/ray-project/release/builds/8522#018dac0b-9111-4b45-89f1-d7a6f83f7d69 for more details.

Managed by OSS Test Policy

### LINKED ISSUE #43192: Release test read_images_train_16_gpu.aws failed
Release test **read_images_train_16_gpu.aws** failed. See https://buildkite.com/ray-project/release/builds/8522#018dabc9-0c01-4639-8a39-ca5587d4e3f9 for more details.

Managed by OSS Test Policy

### LINKED ISSUE #43193: Release test read_images_train_16_gpu_preserve_order.aws failed
Release test **read_images_train_16_gpu_preserve_order.aws** failed. See https://buildkite.com/ray-project/release/builds/8522#018dabcb-3942-4af1-a522-19f01e92bbab for more details.

Managed by OSS Test Policy

### LINKED ISSUE #43194: Release test read_images_train_4_gpu_worker_chaos.aws failed
Release test **read_images_train_4_gpu_worker_chaos.aws** failed. See https://buildkite.com/ray-project/release/builds/8522#018dac0f-be25-47ce-96a0-5f2243baeae0 for more details.

Managed by OSS Test Policy

### LINKED ISSUE #43195: Release test read_images_train_4_gpu_node_chaos.aws failed
Release test **read_images_train_4_gpu_node_chaos.aws** failed. See https://buildkite.com/ray-project/release/builds/8522#018dac10-66c4-44e6-9383-f2b075f093d8 for more details.

Managed by OSS Test Policy

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

https://github.com/ray-project/ray/pull/43146 re-introduced a problem that was originally fixed by https://github.com/ray-project/ray/pull/28826.

The problem:
* When there are large objects captured in the scope of a `Trainer`'s arguments, **the `Trainer` instance on the driver needs to be sure not to package itself (`self`) when registering the actor class** that will be run remotely.
* Otherwise, `self` gets packaged along with the actor class during serialization, and **so do these large objects.**
* Because Ray Train relies on creating a Ray Tune Trainable under the hood to schedule the remote `Trainer` instance (which serves as the dist. training coordinator), and because Ray Tune uses the Ray GCS redis key value store to save a shared copy of the serialized actor class, a huge actor class causes a gRPC error such as the one below:
```
E               ray.tune.error.TuneError: The Trainable/training function is too large for grpc resource limit. Check that its definition is not implicitly capturing a large array or other object in scope. Tip: use tune.with_parameters() to put large objects in the Ray object store. 
E               Original exception: Traceback (most recent call last):
E                 File "/Users/justin/Developer/justinvyu-dev/python/ray/tune/experiment/experiment.py", line 149, in __init__
E                   self._run_identifier = Experiment.register_if_needed(run)
E                 File "/Users/justin/Developer/justinvyu-dev/python/ray/tune/experiment/experiment.py", line 350, in register_if_needed
E                   register_trainable(name, run_object)
E                 File "/Users/justin/Developer/justinvyu-dev/python/ray/tune/registry.py", line 112, in register_trainable
E                   _global_registry.register(TRAINABLE_CLASS, name, trainable)
E                 File "/Users/justin/Developer/justinvyu-dev/python/ray/tune/registry.py", line 239, in register
E                   self.flush_values()
E                 File "/Users/justin/Developer/justinvyu-dev/python/ray/tune/registry.py", line 277, in flush_values
E                   _internal_kv_put(
E                 File "/Users/justin/Developer/justinvyu-dev/python/ray/_private/client_mode_hook.py", line 103, in wrapper
E                   return func(*args, **kwargs)
E                 File "/Users/justin/Developer/justinvyu-dev/python/ray/experimental/internal_kv.py", line 94, in _internal_kv_put
E                   return global_gcs_client.internal_kv_put(key, value, overwrite, namespace) == 0
E                 File "python/ray/_raylet.pyx", line 2614, in ray._raylet._auto_reconnect.wrapper
E                 File "python/ray/_raylet.pyx", line 2591, in ray._raylet._auto_reconnect.wrapper
E                 File "python/ray/_raylet.pyx", line 2728, in ray._raylet.GcsClient.internal_kv_put
E                 File "python/ray/_raylet.pyx", line 579, in ray._raylet.check_status
E               ray.exceptions.RpcError: Sent message larger than max (800008883 vs. 536870912)
```

The fix is to remove the reference to `self` in the train coordinator function trainable. This PR also restructures the code a bit to make it harder to make such a mistake in the future.

### Future considerations

* This redis key value store logic in Tune is probably not necessary and can be re-worked to remove this failure case.

## Related issue number

Closes https://github.com/ray-project/ray/issues/43191
Closes https://github.com/ray-project/ray/issues/43192
Closes https://github.com/ray-project/ray/issues/43193
Closes https://github.com/ray-project/ray/issues/43194
Closes https://github.com/ray-project/ray/issues/43195

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/base_trainer.py [hunk 1]
```diff
@@ -656,38 +691,17 @@ def _generate_trainable_cls(self) -> Type["Trainable"]:
...
-            if checkpoint:
-                # Set `starting_checkpoint` for auto-recovery fault-tolerance
-                # as well as manual restoration.
-                trainer.starting_checkpoint = checkpoint
-            # else: Train will restore from the user-provided
-            # `resume_from_checkpoint` == `starting_checkpoint`.
-
-            # Evaluate datasets if they are wrapped in a factory.
-            trainer.datasets = {
-                k: d() if callable(d) else d for k, d in self.datasets.items()
```
Comment (woshiyyya) [inline python/ray/train/base_trainer.py:679]: Is this the self reference that caused serialization of Huge BaseTrainer object?

**Review by woshiyyya — COMMENTED** [review]
LGTM! One question: What's the change that caused the regression?

File: python/ray/train/base_trainer.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/base_trainer.py:679]: Yep that's it!

**Comment by justinvyu** [comment]
Running release test to verify fix: https://buildkite.com/ray-project/release/builds/8691

**Review by matthewdeng — APPROVED** [review]
<div><br/>via <a href="https://giphy.com/giphystudios2016/">Studios 2016</a> on <a href="https://giphy.com/gifs/studiosoriginals-hand-kiss-fingers-chef-3o7qDWp7hxhi1N8oF2">GIPHY</a></div>

**Pull request merged**